File: dynamic-array/dynamicarray.lua
local function dynamicArray(capacity)
	local a = {
		storage = {},
		capacity = capacity,
		size = 0,
	}

	a.get = function(index)
		return a.storage[index]
	end

	a.set = function(index, value)
		a.storage[index] = value
	end

	a.getSize = function()
		return a.size
	end

	a.getCapacity = function()
		return a.capacity
	end

	a.resize = function()
		a.capacity = a.capacity * 2

		local newStorage = {}
		for i = 1, a.size do
			newStorage[i] = a.storage[i]
		end

		a.storage = newStorage
	end

	a.pushBack = function(value)
		if a.size == a.capacity then
			a.resize()
		end

		a.storage[a.size] = value
		a.size = a.size + 1
	end

	a.popBack = function()
		if a.size > 0 then
			a.size = a.size - 1
		end

		return a.storage[a.size]
	end

	return a
end

return dynamicArray

File: dynamic-array/dynamicarray.go
package main

type DynamicArray struct {
	size     int
	capacity int
	storage  []int
}

func (da *DynamicArray) Get(index int) int {
	return da.storage[index]
}

func (da *DynamicArray) Set(index, value int) {
	da.storage[index] = value
}

func (da *DynamicArray) GetSize() int {
	return da.size
}

func (da *DynamicArray) GetCapacity() int {
	return da.capacity
}

func (da *DynamicArray) PushBack(value int) {
	if da.size == da.capacity {
		da.Resize()
	}
	da.storage[da.size] = value
	da.size++
}

func (da *DynamicArray) PopBack() int {
	if da.size > 0 {
		da.size--
	}
	return da.storage[da.size]
}

func (da *DynamicArray) Resize() {
	da.capacity *= 2
	newStorage := make([]int, da.capacity)

	for i := 0; i < da.size; i++ {
		newStorage = append(newStorage, da.storage[i])
	}

	da.storage = newStorage
}

func NewDynamicArray(capacity int) *DynamicArray {
	return &DynamicArray{
		size:     0,
		capacity: capacity,
		storage:  make([]int, capacity),
	}
}
